cpp/ql/test/query-tests/Security/CWE/CWE-611/tests3.cpp

```diff
@@ -35,14 +35,14 @@ void test3_2(InputSource &data) {
 SAX2XMLReader *p_3_3 = XMLReaderFactory::createXMLReader();
 
 void test3_3(InputSource &data) {
-	p_3_3->parse(data); // BAD (parser not correctly configured) [NOT DETECTED]
+	p_3_3->parse(data); // BAD (parser not correctly configured)
 }
 
 SAX2XMLReader *p_3_4 = XMLReaderFactory::createXMLReader();
 
 void test3_4(InputSource &data) {
 	p_3_4->setFeature(XMLUni::fgXercesDisableDefaultEntityResolution, true);
-	p_3_4->parse(data); // GOOD
+	p_3_4->parse(data); // GOOD [FALSE POSITIVE]
```

Comment:
It looks like barriers aren't working for global variables?  I'm not 100% sure whether the problem lies in IR data flow or in the way this query defines its barriers.

I don't think we should tolerate these false positives (because they are likely to be more common than new good results in the wild), but if you can't easily fix them in the IR dataflow I will have a look at the query, and possibly just block results involving global variables for now.  That will undo all of the changes for this query, but keep the (valuable) improvements in the other queries.

Let me know if you think there's an IR dataflow fix first.

Reply:
I _think_ the issue might be in the way we define transformers using GVN in `DomConfigurationSetParameterTransformer`. It's a totally reasonable way to do it, but since we don't try to do any alias analysis for global variables, the GVNs we get for uses of global variables will pretty much always be unique, which makes GVN for these values quite useless (please correct me if I'm misremembering what the IR actually looks like for uses of global variables @rdmarsh2).

I don't think we have any good way of doing GVN for global variables, so
I think the best solution is to disable detection in the case of global variables in the problematic XML libraries (I guess it's really all of them?).

If we had use-use dataflow (😭) we could use that instead of GVN, and that would support global variables as well.

Reply:
I've created https://github.com/rdmarsh2/ql/pull/5/files as a workaround, but I'm open to suggestions if we can do better.

Reply:
The way this is implemented now, every read of a global variable has flow from every write to that variable (via the variable node). I think that's what's causing the issue here - the initial object creation *also* flows to the use. I'm not sure I can fix that with a reasonable amount of work, though...